### ISSUE TITLE
Flash Attention：从 GPU 内存层次理解注意力加速

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#238]
# Flash Attention：从 GPU 内存层次理解注意力加速

> 不改数学，只改实现——一个 IO-aware 的精确注意力算法

## 背景

Transformer 的自注意力机制是 O(N²) 的计算和存储开销。但真正的瓶颈不在计算量——现代 GPU 的算力增长远快于显存带宽增长。实际卡脖子的是数据搬运：中间结果在显存和计算单元之间反复传输，计算核心大部分时间在等数据。

Flash Attention（Tri Dao, 2022）的核心洞察就是：**Attention 是 memory-bound 的，不是 compute-bound 的**。既然瓶颈在 IO，那就从 IO 入手优化。

---

## 一、GPU 内存层次

要理解 Flash Attention 为什么有效，先要理解 GPU 的内存结构。

### 1.1 两层关键存储

| 层级 | 名称 | 技术 | 容量 | 带宽（H100） |
|---|---|---|---|---|
| 片上 | SRAM（Shared Memory / L2 Cache） | 静态 RAM | ~50 MB | ~19 TB/s |
| 片外 | HBM（主显存） | 动态 RAM 堆叠 | 80 GB | ~3.35 TB/s |

核心数字：**SRAM 带宽是 HBM 的 6 倍，但容量只有 HBM 的 1/1600**。

### 1.2 SRAM vs DRAM

这两层存储的底层技术完全不同：

| | SRAM（片上） | DRAM（片外，HBM 的底层） |
|---|---|---|
| 存储单元 | 6 个晶体管组成的锁存器 | 1 个晶体管 + 1 个电容 |
| 是否需要刷新 | 不需要（"静态"之名由来） | 需要周期性刷新电荷 |
| 速度 | 快 | 慢 |
| 密度 | 低（面积大） | 高（紧凑） |
| 成本 | 贵 | 便宜 |

SRAM 直接制造在 GPU die 上，紧挨计算单元，所以访问延迟极低。但 6T 结构占面积大，不可能做到 GB 级。

DRAM 用电容存电荷表示 0/1，结构紧凑但需要持续刷新（"动态"之名由来），速度也更慢。

### 1.3 HBM 是什么

HBM（High Bandwidth Memory）不是一种新的存储技术，而是 DRAM 的一种**高带宽封装形态**：

```
DRAM（底层技术）
├── DDR（DDR4, DDR5）         → CPU 主存
├── GDDR（GDDR6, GDDR6X）    → 消费级 GPU（RTX 4090）
└── HBM（HBM2e, HBM3, HBM3e）→ 数据中心 GPU（A100, H100）
```

HBM 的做法是把多层 DRAM die 垂直堆叠（3D stacking），层间用硅通孔（TSV）连接，通过硅中介层（interposer）与 GPU 紧密耦合。接口位宽极大（1024-bit），总带宽拉到 TB/s 级别。代价是容量受限、成本高、散热难。

---

## 二、标准 Attention 的 IO 问题

标准自注意力的朴素实现：

```python
S = Q @ K.T        # [N, N]  写入 HBM
P = softmax(S)     # [N, N]  读 S，写 P 到 HBM
O = P @ V          # [N, d]  读 P，写 O 到 HBM
```

问题在哪：

1. **物化完整的 N×N 矩阵**：S 和 P 都是 N×N 大小，序列长度 N=4096 时，仅 S 矩阵就占 64 MB（FP32），必须写入 HBM
2. **三次 HBM 往返**：计算 S 写出去，读回来算 softmax，P 写出去，读回来乘 V——每一步都要经历 HBM 的高延迟
3. **显存 O(N²)**：序列一长就 OOM

GPU 的计算单元大部分时间不是在算，而是在等 HBM 把数据送过来。这就是 memory-bound。

---

### 3.1 核心策略：Tiling + Kernel Fusion

既然 SRAM 快但小，HBM 慢但大，策略就是：

> **把 Q/K/V 分成小块，每块加载到 SRAM，在 SRAM 内完成所有计算，只把最终结果写回 HBM。永远不物化完整的 N×N 矩阵。**

这是用"多算几次"（FLOPs 便宜）换"少搬几次"（IO 昂贵）。

### 3.2 关键难点：分块 Softmax

Tiling 的想法直觉上简单，但 softmax 天然是全局操作——需要所有元素的 max 和 sum 来归一化。如何在只看到部分数据时计算 softmax？

解法是 **Online Softmax**（在线 softmax）：

```
对每个新块：
1. 计算当前块的局部 max
2. 用 running max 修正之前累积的 sum 和 output
3. 更新全局 max 和 sum
```

通过维护 running max 和 rescaling 系数，分块计算的最终结果与全局 softmax 数学上完全等价。不是近似，是精确的。

### 3.3 算法概要

```
for each block of Q:
    for each block of K, V:
        1. 从 HBM 加载 Q_block, K_block, V_block 到 SRAM
        2. 在 SRAM 中计算 S_block = Q_block @ K_block.T
        3. 在 SRAM 中计算局部 softmax（带 online rescaling）
        4. 在 SRAM 中累积 O_block += P_block @ V_block
    写回最终的 O_block 到 HBM
```

整个过程中，HBM 只被读写 O(N) 量级的数据（Q、K、V 和最终输出 O），N×N 的中间矩阵始终只在 SRAM 中以块的形式存在。

### 3.4 收益

| 指标 | 标准 Attention | Flash Attention |
|---|---|---|
| HBM 读写量 | O(N² + Nd) | O(N²d² / M)，M 为 SRAM 大小 |
| 显存占用 | O(N²) | O(N) |
| 实测加速 | 基线 | 2-4x |

---

## 四、与"高效注意力"的区别

在 Flash Attention 之前，社区有大量"高效注意力"研究：

| 方法 | 思路 | 牺牲精度？ |
|---|---|---|
| Sparse Attention | 只计算部分位置的注意力 | 是 |
| Linear Attention | 用核函数近似 softmax | 是 |
| Performer | 随机特征近似注意力矩阵 | 是 |
| Linformer | 低秩投影压缩 K、V | 是 |
| **Flash Attention** | **IO 层面优化，不改计算逻辑** | **否** |

Flash Attention 的定位清晰：**不改数学，只改实现**。它输出的结果与标准 attention 逐位相同（在浮点误差内）。这意味着它可以无缝替换任何使用标准 attention 的模型，无需重新训练或验证。

---

### Flash Attention 1（2022）

奠定基础：tiling + online softmax + kernel fusion。在 A100 上实测 2-4x 加速。

### Flash Attention 2（2023）

在算法不变的前提下优化 GPU 利用率：
- 减少非 matmul 的 FLOPs（优化 rescaling 计算）
- 优化 warp 之间的工作分配，减少通信
- 更好的并行度——同时在序列长度和 batch 维度并行

结果：在 Flash Attention 1 基础上再提升约 2x，接近 H100 理论峰值吞吐的 70%。

### Flash Attention 3（2024）

针对 Hopper 架构（H100/H200）的硬件特性深度优化：
- **TMA**（Tensor Memory Accelerator）：硬件级异步数据搬运
- **WGMMA**（Warp Group Matrix Multiply-Accumulate）：新的矩阵运算指令
- **FP8 支持**：利用 FP8 tensor core 进一步提升吞吐
- **计算与数据搬运重叠**：Hopper 的异步特性允许搬下一块数据的同时计算当前块

---

## 六、工程影响

Flash Attention 现在是 Transformer 训练和推理的标配基础设施：

- **PyTorch 2.0+** 内置 `torch.nn.functional.scaled_dot_product_attention`，自动调用 Flash Attention kernel
- **主流框架全面支持**：Hugging Face Transformers、vLLM、TensorRT-LLM 等
- **长序列成为可能**：显存从 O(N²) 降到 O(N)，直接解锁了 32K、128K 甚至更长的上下文窗口

没有 Flash Attention，今天的长上下文模型（GPT-4 128K、Claude 200K）在工程上很难落地。

---

## 总结

Flash Attention 的核心洞察可以一句话概括：

> **现代 GPU 上，搬数据比算数据贵。把计算搬到数据旁边（SRAM），而不是把数据搬到计算旁边。**

它不是算法创新（注意力的数学没变），而是系统层面的工程创新——深入理解硬件特性后，重新组织计算顺序来最大化 IO 效率。这种"hardware-aware algorithm design"的思路，在 AI 系统优化中越来越重要。

---

## 参考资料

- Dao et al., *FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness*, NeurIPS 2022
- Dao, *FlashAttention-2: Faster Attention with Better Parallelism and Work Partitioning*, ICLR 2024
- Shah et al., *FlashAttention-3: Fast and Accurate Attention with Asynchrony and Low-precision*, 2024
- Milakov & Gimelshein, *Online Normalizer Calculation for Softmax*, 2018
- https://github.com/Dao-AILab/flash-attention
